src/main/kotlin/com/sourcegraph/cody/edit/FixupSession.kt

```diff
@@ -0,0 +1,354 @@
+    ApplicationManager.getApplication().invokeLater {
+      val logicalPosition = LogicalPosition(range.start.line, range.start.character)
+      editor.scrollingModel.scrollTo(logicalPosition, ScrollType.CENTER)
+    }
+  }
+
+  private fun showWorkingGroup() {
+    showLensGroup(LensGroupFactory(this).createTaskWorkingGroup())
+  }
+
+  private fun showAcceptGroup() {
+    showLensGroup(LensGroupFactory(this).createAcceptGroup())
+  }
+
+  fun finish() {
+    try {
+      controller.removeSession(this)
+      rangeMarkers.forEach { it.dispose() }
+      rangeMarkers.clear()
+    } catch (x: Exception) {
+      logger.debug("Session cleanup error", x)
+    }
+    Disposer.dispose(this)
+  }
+
+  /** Subclass sends a fixup command to the agent, and returns the initial task. */
+  abstract fun makeEditingRequest(agent: CodyAgent): CompletableFuture<EditTask>
+
+  fun accept() {
+    CodyAgentService.withAgent(project) { agent ->
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_ACCEPT, listOf(taskId!!)))
+    }
+    finish()
+  }
+
+  fun cancel() {
+    CodyAgentService.withAgent(project) { agent ->
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_CANCEL, listOf(taskId!!)))
+    }
+    if (performedEdits) {
+      undo()
+    } else {
+      finish()
+    }
+  }
+
+  abstract fun retry()
+
+  abstract fun diff()
+
+  fun undo() {
+    CodyAgentService.withAgent(project) { agent ->
+      agent.server.commandExecute(CommandExecuteParams(COMMAND_UNDO, listOf(taskId!!)))
+    }
+    undoEdits()
+    finish()
+  }
+
+  fun performInlineEdits(edits: List<TextEdit>) {
+    val project = editor.project ?: return
+    val doc: Document = editor.document
+
+    // TODO: This is an artifact of the update to concurrent editing tasks.
+    // We do need to mute any LensGroup listeners, but this is an ugly way to do it.
+    // There are multiple Lens groups; we need a Document-level listener list.
+    lensGroup?.withListenersMuted {
+      if (!controller.isEligibleForInlineEdit(editor)) {
+        return@withListenersMuted logger.warn("Inline edit not eligible")
+      }
+      // Mark all the edit locations so the markers will move as we edit the document,
+      // preserving the original positions of the edits.
+      val markers = edits.mapNotNull { createMarkerForEdit(doc, it) }
+      val sortedEdits = edits.zip(markers).sortedByDescending { it.second.startOffset }
+      // Apply the edits in a write action.
+      WriteCommandAction.runWriteCommandAction(project) {
+        for ((edit, marker) in sortedEdits) {
+          when (edit.type) {
+            "replace",
+            "delete" -> ReplaceUndoableAction(this, edit, marker)
+            "insert" -> InsertUndoableAction(this, edit, marker)
+            else -> logger.warn("Unknown edit type: ${edit.type}")
+          }
+        }
+      }
+    }
+  }
+
+  private fun createMarkerForEdit(doc: Document, edit: TextEdit): RangeMarker? {
```

Comment:
It seems to me that logic from there could be just moved inside `ReplaceUndoableAction` and `InsertUndoableAction`, and you wouldn't have to pass marker through the constructor.